File: java/exercises/matrix_multiplier.java
/*
 *	MIT License
 *
 *	Copyright (c) 2016 Johnathan Fercher
 *
 *	Permission is hereby granted, free of charge, to any person obtaining a copy
 *	of this software and associated documentation files (the "Software"), to deal
 *	in the Software without restriction, including without limitation the rights
 *	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *	copies of the Software, and to permit persons to whom the Software is
 *	furnished to do so, subject to the following conditions:
 *
 *	The above copyright notice and this permission notice shall be included in all
 *	copies or substantial portions of the Software.
**/

import java.util.Random;

//! class which defines multiplication of matrices 
public class matrix_multiplier {
    private static int rows, cols;                          //! number of lines and columns
    private static float [][]A;                             
    private static float [][]B;
    private static float [][]R;

    private static Random randomGenerator;

    //! method which fill a matrix with random values
    public static void fill_matrices(){
        for(int i = 0 ; i < rows; i++){
            for(int j = 0 ; j < cols ; j++){
                A[i][j] = randomGenerator.nextInt(10);
                B[i][j] = randomGenerator.nextInt(10);
            }
        }
    }

    //! function which multiply two matrix
    public static void multiply_matrices(){
        float hold = 0;

        for(int i = 0 ; i < rows ; i++){
            for(int j = 0 ; j < cols ; j++){
                hold = 0;

                for(int k = 0 ; k < rows ; k++){
                    hold += A[i][k] * B[k][j];
                }

                R[i][j] = hold;
            } 
        }
    }

    //! method which prints a matrix
    public static void print_matrices(){
        for(int i = 0 ; i < rows; i++){
            for(int j = 0 ; j < cols ; j++){
                System.out.print("\t" + A[i][j]);
            }
            System.out.println("");
        }

        for(int i = 0 ; i < rows; i++){
            for(int j = 0 ; j < cols ; j++){
                System.out.print("\t" + B[i][j]);
            }
            System.out.println("");
        }

        for(int i = 0 ; i < rows; i++){
            for(int j = 0 ; j < cols ; j++){
                System.out.print("\t" + R[i][j]);
            }
            System.out.println("");
        }
    }

    public static void main(String[] args) {
        rows = Integer.parseInt(args[0]);
        cols = Integer.parseInt(args[1]);    

        A = new float[rows][cols];
        B = new float[cols][rows];
        R = new float[rows][rows];

        randomGenerator = new Random();

        fill_matrices();
        multiply_matrices();
        //print_matrices();
    }
}

File: javascript/exercises/matrix_multiplier.js
/*
 *	MIT License
 *
 *	Copyright (c) 2016 Rafael Almeida
 *
 *	Permission is hereby granted, free of charge, to any person obtaining a copy
 *	of this software and associated documentation files (the "Software"), to deal
 *	in the Software without restriction, including without limitation the rights
 *	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *	copies of the Software, and to permit persons to whom the Software is
 *	furnished to do so, subject to the following conditions:
 *
 *	The above copyright notice and this permission notice shall be included in all
 *	copies or substantial portions of the Software.
**/

//! function which allocates a matrix dynamically in memory
function alloc_matrix(rows, cols) {
	var m = [];
	for (var i = 0; i < rows; i++) {
		m.push([]);
		for (var j = 0; j < cols; j++) {
			m[i][j] = Math.floor(Math.random()  * 10);
		}
	}
	return m;
}

//! function which prints a matrix
function print_matrix(matrix) {
	var str;
	for (var i = 0; i < matrix.length; i++) {
		str = "";
		for (var j = 0; j < matrix[i].length; j++) {
			str += matrix[i][j] + " ";
		}
		document.write(str + '<br>');
	}
	document.write('<br>');
	return matrix;
}

//! function which multiply two matrix
function multiply_matrix(a, b) {
	var r = alloc_matrix(a.length, b[0].length);
	for (var i = 0; i < a.length; i++) {
		for (var j = 0; j < b[0].length; j++) {
			var hold = 0;
			for (var k = 0; k < a[0].length; k++) {
				hold = hold + a[i][k] * b[k][j];
			}
			r[i][j] = hold;
		}
	}
	return r;
}

var i = parseInt(process.argv[2]);
var j = parseInt(process.argv[3]);

var matrixA = alloc_matrix(i, j);
var matrixB = alloc_matrix(j, i);
var matrixR = multiply_matrix(matrixA, matrixB);

//print_matrix(matrixA);
//print_matrix(matrixB);
//print_matrix(matrixR);